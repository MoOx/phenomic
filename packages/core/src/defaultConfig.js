import path from "path";

export default {
  path: process.cwd(),
  outdir: path.join(process.cwd(), "dist"),
  port: 3333,
  bundleName: "phenomic.main",
  plugins: []
};
