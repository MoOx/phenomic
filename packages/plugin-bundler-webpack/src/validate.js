import multili from "multili";

const typeOf = value => {
  const type = typeof value;
  if (Array.isArray(value)) {
    return "array";
  }
  if (value === null) {
    return "null";
  }
  return type;
};

export default (webpackConfig: Object, config: PhenomicConfig) => {
  if (!webpackConfig.entry || !webpackConfig.entry[config.bundleName]) {
    let error = `@phenomic/plugin-bundler-webpack expects an entry named '${
      config.bundleName
    }' in your webpack config.\n`;
    // be careful about the indentation here
    let adviceCode = `[
            process.env.PHENOMIC_ENV !== "static" && require.resolve("webpack-hot-middleware/client"),
            process.env.PHENOMIC_ENV !== "static" && require.resolve("react-hot-loader/patch"),
            "./App.js"
          ].filter(item => item)`;
    const type = typeOf(webpackConfig.entry);
    if (type === "object") {
      // $FlowFixMe here we know we have an object, see typeOf function
      error += `Your current entries keys are '${Object.keys(
        webpackConfig.entry
      )}'.\n`;
    } else {
      error += `Your current entry is of type '${type}'.\n`;
    }
    if (type === "string" || type === "array") {
      adviceCode = JSON.stringify(webpackConfig.entry);
    }
    error += multili(`
      Please use an object like this:

        entry: {
          "phenomic.main": ${adviceCode},
        }

    `);
    throw new Error(error);
  }
};
