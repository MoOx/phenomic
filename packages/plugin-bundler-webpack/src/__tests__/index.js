import defaultConfig from "@phenomic/core/lib/defaultConfig.js";

import validate from "../validate.js";

const defaultWebpackConfig = require("../webpack.config.js");

test("validate default config", () => {
  expect(() => {
    validate(defaultWebpackConfig(defaultConfig), defaultConfig);
  }).not.toThrow();
});

test("validate nothing", () => {
  try {
    validate({}, defaultConfig);
    throw new Error("pass but should not");
  } catch (e) {
    expect(e).toMatchSnapshot();
  }
});

test("validate string", () => {
  try {
    validate(
      {
        entry: "App.js"
      },
      defaultConfig
    );
    throw new Error("pass but should not");
  } catch (e) {
    expect(e).toMatchSnapshot();
  }
});

test("validate array", () => {
  try {
    validate(
      {
        entry: ["something.js", "App.js"]
      },
      defaultConfig
    );
    throw new Error("pass but should not");
  } catch (e) {
    expect(e).toMatchSnapshot();
  }
});

test("validate object without key", () => {
  try {
    validate(
      {
        entry: {
          stuff: ["something.js", "App.js"],
          thing: ["something.js", "App.js"]
        }
      },
      defaultConfig
    );
    throw new Error("pass but should not");
  } catch (e) {
    expect(e).toMatchSnapshot();
  }
});
