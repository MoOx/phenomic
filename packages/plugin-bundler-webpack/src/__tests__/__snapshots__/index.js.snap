// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validate array 1`] = `
[Error: @phenomic/plugin-bundler-webpack expects an entry named 'phenomic.main' in your webpack config.
Your current entry is of type 'array'.

Please use an object like this:

  entry: {
    "phenomic.main": ["something.js","App.js"],
  }

]
`;

exports[`validate nothing 1`] = `
[Error: @phenomic/plugin-bundler-webpack expects an entry named 'phenomic.main' in your webpack config.
Your current entry is of type 'undefined'.

Please use an object like this:

  entry: {
    "phenomic.main": [
      process.env.PHENOMIC_ENV !== "static" && require.resolve("webpack-hot-middleware/client"),
      process.env.PHENOMIC_ENV !== "static" && require.resolve("react-hot-loader/patch"),
      "./App.js"
    ].filter(item => item),
  }

]
`;

exports[`validate object without key 1`] = `
[Error: @phenomic/plugin-bundler-webpack expects an entry named 'phenomic.main' in your webpack config.
Your current entries keys are 'stuff,thing'.

Please use an object like this:

  entry: {
    "phenomic.main": [
      process.env.PHENOMIC_ENV !== "static" && require.resolve("webpack-hot-middleware/client"),
      process.env.PHENOMIC_ENV !== "static" && require.resolve("react-hot-loader/patch"),
      "./App.js"
    ].filter(item => item),
  }

]
`;

exports[`validate string 1`] = `
[Error: @phenomic/plugin-bundler-webpack expects an entry named 'phenomic.main' in your webpack config.
Your current entry is of type 'string'.

Please use an object like this:

  entry: {
    "phenomic.main": "App.js",
  }

]
`;
